improver/utilities/gradient_between_vertical_levels.py

```diff
@@ -0,0 +1,150 @@
+        height coordinate. If the cubes are provided at pressure levels, the height above sea level
+        is extracted from a geopotential_height cube.If both cubes have a
+        height coordinate then no additional cubes are required.
+
+        Args:
+            cubes:
+                Two cubes containing a diagnostic at two different vertical levels. The cubes must
+                contain either a height or pressure scalar coordinate. If the cubes contain a height
+                scalar coordinate this is assumed to be a height above ground level.
+            geopotential_height:
+                Optional cube that contains the height above sea level of pressure levels. This cube
+                is required if any input cube is defined at pressure level. This is used to extract
+                the height above sea level at the pressure level of the input cubes.
+            orography:
+                Optional cube containing the orography height above sea level. This cube is required
+                if exactly one input cube is defined at height levels and is used to convert the
+                height above ground level to height above sea level.
+
+        Returns:
+            A cube containing the gradient between the cubes between two vertical levels.
+
+        Raises:
+            ValueError: If exactly one input cube is defined at height levels and no orography cube
+                        is provided.
+            ValueError: If either input cube is defined at pressure levels and no
+                        geopotential_height cube is provided.
+        """
+
+        cube_heights = []
+
+        for cube in cubes:
+            try:
+                cube_height = np.array(cube.coord("height").points)
+            except CoordinateNotFoundError:
+                if geopotential_height:
+                    height_ASL = geopotential_height.extract(
+                        iris.Constraint(pressure=cube.coord("pressure").points)
+                    )
+                else:
+                    raise ValueError(
+                        """No geopotential height cube provided but one of the inputs cubes has a
+                        pressure coordinate"""
+                    )
+            else:
+                if orography:
+                    height_ASL = orography + cube_height
+                elif not (orography or geopotential_height):
+                    height_ASL = cube_height
+                else:
+                    raise ValueError(
+                        """No orography cube provided but one of the input cubes has height
+                        coordinate"""
+                    )
+            cube_heights.append(height_ASL)
+
+        height_diff = cube_heights[0] - cube_heights[1]
+        height_diff.data = np.ma.masked_where(height_diff.data == 0, height_diff.data)
+
+        diff = cubes[0] - cubes[1]
+        gradient = diff / height_diff
```

Comment:
What do the output metadata look like if the two source cubes represent different times? Should we detect and reject such cases?

Reply:
I've added in a line to test the cubes time coordinates match. We don't have a use case for a gradient between different time steps but its not hard to change in future if we need to.